modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectDefinitionVulcanBatchEngineTaskItemDelegate.java

```diff
@@ -0,0 +1,265 @@
+					"company would be used instead.");
+		}
+	}
+
+	@Override
+	public void setContextUser(User contextUser) {
+		_contextUser = contextUser;
+	}
+
+	@Override
+	public void setLanguageId(String languageId) {
+		_contextAcceptLanguage = new AcceptLanguage() {
+
+			@Override
+			public List<Locale> getLocales() {
+				return null;
+			}
+
+			@Override
+			public String getPreferredLanguageId() {
+				return languageId;
+			}
+
+			@Override
+			public Locale getPreferredLocale() {
+				return LocaleUtil.fromLanguageId(languageId);
+			}
+
+		};
+	}
+
+	@Override
+	public void update(
+			Collection<ObjectDefinition> items,
+			Map<String, Serializable> parameters)
+		throws Exception {
+
+		throw new UnsupportedOperationException();
+	}
+
+	protected <T, R, E extends Throwable> List<R> transformToList(
+		T[] array, UnsafeFunction<T, R, E> unsafeFunction) {
+
+		return TransformUtil.transformToList(array, unsafeFunction);
+	}
+
+	private com.liferay.object.model.ObjectDefinition _create(
+			ObjectDefinition objectDefinition)
+		throws Exception {
+
+		return _objectDefinitionService.addCustomObjectDefinition(
+			LocalizedMapUtil.getLocalizedMap(objectDefinition.getLabel()),
+			objectDefinition.getName(), objectDefinition.getPanelAppOrder(),
+			objectDefinition.getPanelCategoryKey(),
+			LocalizedMapUtil.getLocalizedMap(objectDefinition.getPluralLabel()),
+			objectDefinition.getScope(), objectDefinition.getStorageType(),
+			transformToList(
+				objectDefinition.getObjectFields(),
+				objectField -> ObjectFieldUtil.toObjectField(
+					objectField, _objectFieldLocalService,
+					_objectFieldSettingLocalService,
+					_objectFilterLocalService)));
+	}
+
+	private void _createWithWorkflow(ObjectDefinition objectDefinition)
+		throws Exception {
+
+		com.liferay.object.model.ObjectDefinition objectDefinition1 = _create(
+			objectDefinition);
+
+		Status status = objectDefinition.getStatus();
+
+		if (status.getCode() == WorkflowConstants.STATUS_APPROVED) {
+			_objectDefinitionService.publishCustomObjectDefinition(
+				objectDefinition1.getObjectDefinitionId());
+		}
+	}
+
+	private UnsafeConsumer<ObjectDefinition, Exception> _getUnsafeConsumer(
+		Map<String, Serializable> parameters) {
+
+		Boolean applyWorkflowStatus = (Boolean)parameters.getOrDefault(
+			"applyWorkflowStatus", Boolean.TRUE);
+
+		String createStrategy = (String)parameters.getOrDefault(
+			"createStrategy", "INSERT");
+
+		if ("INSERT".equalsIgnoreCase(createStrategy)) {
+			if (!applyWorkflowStatus) {
+				return objectDefinition -> _create(objectDefinition);
+			}
+
+			return objectDefinition -> _createWithWorkflow(objectDefinition);
+		}
+
+		if (!applyWorkflowStatus) {
+			return objectDefinition -> _upsert(objectDefinition);
+		}
+
+		return objectDefinition -> _upsertWithWorkflow(objectDefinition);
+	}
+
+	private void _upsert(ObjectDefinition objectDefinition) {
+
+		// https://issues.liferay.com/browse/LPS-162113
```

Comment:
As I said in the meeting, upsert seems to be implemented in https://github.com/brianchandotcom/liferay-portal/pull/119404/files. We should make sure we don't conflict with those changes by syncing with the objects team.